src/picongpu/include/particles/manipulators/FreeRngImpl.hpp

```diff
@@ -0,0 +1,194 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
+#include "simulation_defines.hpp"
+#include "nvidia/rng/RNG.hpp"
+#include "nvidia/rng/methods/Xor.hpp"
+#include "mpi/SeedPerRank.hpp"
+#include "traits/GetUniqueTypeId.hpp"
+
+#include <utility>
+#include <type_traits>
+
+namespace picongpu
+{
+namespace particles
+{
+namespace manipulators
+{
+
+    template<
+        typename T_Functor,
+        typename T_Distribution,
+        typename T_SpeciesType
+    >
+    struct FreeRngImpl : private T_Functor
+    {
+
+        using Functor = T_Functor;
+        using Distribution = T_Distribution;
+        using SpeciesType = T_SpeciesType;
+        using SpeciesName = typename MakeIdentifier<SpeciesType>::type;
+
+        /** constructor
+         *
+         * This constructor is only compiled if the user functor has
+         * a host side constructor with one (uint32_t) argument.
+         *
+         * @tparam DeferFunctor is used to defer the functor type evaluation to enable/disable
+         *                      the constructor
+         * @param currentStep current simulation time step
+         * @param is used to enable/disable the constructor (do not pass any value to this parameter)
+         */
+        template< typename DeferFunctor = Functor >
+        HINLINE FreeRngImpl(
+            uint32_t currentStep,
+            typename std::enable_if<
+                std::is_constructible<
+                    DeferFunctor,
+                    uint32_t
+                >::value
+            >::type* = 0
+        ) : Functor( currentStep ), isInitialized( false )
+        {
+            hostInit( currentStep );
+        }
+
+        /** constructor
+         *
+         * This constructor is only compiled if the user functor has a default constructor.
+         *
+         * @tparam DeferFunctor is used to defer the functor type evaluation to enable/disable
+         *                      the constructor
+         * @param currentStep simulation time step
+         * @param is used to enable/disable the constructor (do not pass any value to this parameter)
+         */
+        template< typename DeferFunctor = Functor >
+        HINLINE FreeRngImpl(
+            uint32_t currentStep,
+            typename std::enable_if<
+                std::is_constructible< DeferFunctor >::value
+            >::type* = 0
+        ) : Functor( ), isInitialized( false )
+        {
+            hostInit( currentStep );
+        }
+
+        /** call user functor
+         *
+         * The random number generator is initialized with the first call.
+         *
+         * @param cell superCell index within the local volume
+         * @param unused
+         * @param isParticle1 define if the reference @p particleSpecies1 is valid
+         * @param unused
```

Comment:
QUESTION: Why do you need a unused variable? Do to the (required) common interface?